File: refsol/word-ladder.js
export function ladderLength(beginWord, endWord, wordList) {
  const wordSet = new Set(wordList);
  if (!wordSet.has(endWord)) return 0;
  if (beginWord === endWord) return 1;

  const alphabet = "abcdefghijklmnopqrstuvwxyz";
  let queue = [beginWord];
  const visited = new Set([beginWord]);
  let steps = 1;

  while (queue.length) {
    const next = [];
    for (const word of queue) {
      if (word === endWord) return steps;
      for (let i = 0; i < word.length; i++) {
        for (const c of alphabet) {
          if (c === word[i]) continue;
          const candidate = word.slice(0, i) + c + word.slice(i + 1);
          if (wordSet.has(candidate) && !visited.has(candidate)) {
            visited.add(candidate);
            next.push(candidate);
          }
        }
      }
    }
    queue = next;
    steps++;
  }

  return 0;
}

File: refsol/word-ladder.py
from collections import deque


def ladderLength(beginWord, endWord, wordList):
    word_set = set(wordList)
    if endWord not in word_set:
        return 0
    if beginWord == endWord:
        return 1

    queue = deque([(beginWord, 1)])
    visited = {beginWord}

    while queue:
        word, steps = queue.popleft()
        if word == endWord:
            return steps
        for i in range(len(word)):
            for c in "abcdefghijklmnopqrstuvwxyz":
                if c == word[i]:
                    continue
                candidate = word[:i] + c + word[i + 1:]
                if candidate in word_set and candidate not in visited:
                    visited.add(candidate)
                    queue.append((candidate, steps + 1))

    return 0
